fe_utils/solvers/helmholtz.py

```diff
@@ -29,6 +33,52 @@ def assemble(fs, f):
 
     # Now loop over all the cells and assemble A and l
 
+    cell_node_map = fs.cell_nodes
+
+    num_quad_point = len(quad_rule.points)
+
+    num_of_cells = cell_node_map.shape[0]
+
+    for c in range(num_of_cells):
+        # Note with our change of coordinates from cell c to reference cell, jacobian is constant in the cell.
+        # Compute some essential Jacobian related terms.
+        jacobian = fs.mesh.jacobian(c)
+        det_j = np.abs(np.linalg.det(jacobian))
+        J_inv_T = np.transpose(np.linalg.inv(jacobian))
+
+        # Number of nodes in our ref cell.
+        node_count_cell = fs.element.node_count
+
+        # vector storing the values of f at the nodes of cell c
+        f_val_in_cell_c = [f.values[cell_node_map[c, k]] for k in range(node_count_cell)]
+
+        # vector where q^{th} component stores values of f at the q^th quadrature point.
+        integral_f_cell_c = [np.dot(f_val_in_cell_c, basis_at_quad[q, :]) for q in range(num_quad_point)]
+
+        # Constructing the matrix A and l.
+        for i in range(node_count_cell):
+            # value of element in vector l.
+            integral_l = 0
+            # row_index for A and for l.
+            row_index = cell_node_map[c, i]
+            for q in range(num_quad_point):
+                integral_l += basis_at_quad[q, i] * integral_f_cell_c[q] * quad_rule.weights[q]*det_j
+            # update the correct element of l.
+            l[cell_node_map[c, i]] += integral_l
+            for j in range(node_count_cell):
+                integral = 0
+                for q in range(num_quad_point):
+                    # Compute different components of the integral at different quadrature points.
+                    grad_term = np.dot((J_inv_T  @  basis_grad_at_quad[q, i]), (J_inv_T @ basis_grad_at_quad[q, j]))
+                    product_term = basis_at_quad[q, i] * basis_at_quad[q, j]
+                    integral += (grad_term + product_term) * quad_rule.weights[q] * det_j
+                # column index for the entry of A we may change.
+                col_index = cell_node_map[c, j]
+                # Only edit the matrix A if the integral is non-zero.
+                if integral != 0:
+                    A[np.ix_(np.array([row_index]), np.array([col_index]))] += np.array([integral])
```

Comment:
Hand the i, j, and q loops with lots of nested linear algebra calls inside is guaranteed to make this code very slow. It's also hard to read and uses many more lines than is needed. This could all be replaced by a lot fewer lines using `np.einsum`.

The use of `np.ix_` here makes no sense, nor does the casting of scalars to arrays.